scertecd/scertecd.go

```diff
@@ -660,12 +669,165 @@ func (cu *certUpdateCheck) updateIfNeeded(ctx context.Context, prev *certUpdateR
+		return err
+	}
+
+	// enumerate the roots to sign with
+	roots, err := cu.roots(ctx)
+	if err != nil {
+		return err
+	}
+
+	// write the private key to the PEM output
+	var allPEM bytes.Buffer
+	if err := encodePrivateKeyPEM(&allPEM, privKey); err != nil {
+		return err
+	}
+
+	// append signatures from each root to the PEM output
+	for _, r := range roots {
+		derBytes, err := cu.signWithRoot(ctx, r, newCert, privKey)
+		if err != nil {
+			cu.s.addError("private-renewal-sign-with-root")
+			return fmt.Errorf("failed to sign new cert with root %q: %w", r, err)
+		}
+		pb := &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: derBytes,
+		}
+		if err := pem.Encode(&allPEM, pb); err != nil {
+			cu.s.addError("private-renewal-pem-encode")
+			return fmt.Errorf("failed to PEM encode cert signed by root %q: %w", r, err)
+		}
+	}
+
+	m, err := cu.s.parseCertMeta(allPEM.Bytes(), true)
+	if err != nil {
+		cu.s.addError("private-renewal-parse-new-cert")
+		return fmt.Errorf("failed to parse newly fetched cert: %w", err)
+	}
+
+	cu.res.SecretVersion, err = cu.s.putAndActivateSecret(ctx, cu.Logf, secName, allPEM.Bytes())
+	if err != nil {
+		return err
+	}
+	cu.res.Updated = true
+	cu.res.CertMeta = m
+	return nil
+}
+
+// signWithRoot signs newCert with the root CA identified by root, using newKey
+// as the private key for newCert.
+func (cu *certUpdateCheck) signWithRoot(ctx context.Context, root string, newCert *x509.Certificate, newKey crypto.Signer) ([]byte, error) {
+	// retrieve root private key; PEM decode & parse it
+	caPrivKeyName := path.Join(cu.s.Prefix, "roots", root, "private-key")
+	caPrivKeySecret, err := cu.s.getSecret(ctx, caPrivKeyName)
+	if err != nil {
+		return nil, fmt.Errorf("getSecret %q: %w", caPrivKeyName, err)
+	}
+	keyPB, _ := pem.Decode(caPrivKeySecret.Value)
+	if keyPB == nil {
+		return nil, fmt.Errorf("pem.Decode %q: no PEM data found", caPrivKeyName)
+	}
+	caPrivKey, err := x509.ParsePKCS8PrivateKey(keyPB.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse private key %q: %v", caPrivKeyName, err)
+	}
+
+	// retrieve root certificate; PEM decode & parse it
+	caCertName := path.Join(cu.s.Prefix, "roots", root, "certificate")
+	caCertSecret, err := cu.s.getSecret(ctx, caCertName)
+	if err != nil {
+		return nil, fmt.Errorf("getSecret %q: %w", caCertName, err)
+	}
+	certPB, _ := pem.Decode(caCertSecret.Value)
+	if certPB == nil {
+		return nil, fmt.Errorf("pem.Decode %q: no PEM data found", caCertName)
+	}
+	caCert, err := x509.ParseCertificate(certPB.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("parsePEMCertificates %q: %w", caCertName, err)
+	}
+
+	// sign the new certificate with the root private key
+	return x509.CreateCertificate(rand.Reader, newCert, caCert, newKey.Public(), caPrivKey)
+}
+
+// roots returns a list of root CA keys found in setec under the "roots/" prefix.
+func (cu *certUpdateCheck) roots(ctx context.Context) ([]string, error) {
+	l, err := cu.s.SetecClient.List(ctx)
```

Comment:
Try to find a better variable name than l? https://docs.google.com/document/d/1ooRdYeSYI6VdxE4rOFULdkeN8WRG1SMLqVV4YByz_aU/edit?tab=t.0#heading=h.q6x7cl615qqs